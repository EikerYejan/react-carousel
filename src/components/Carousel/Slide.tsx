import styled from 'styled-components'

type Props = React.HTMLAttributes<HTMLDivElement> & {
  children: React.ReactNode
}

const Wrapper = styled.div`
  height: 100%;
  display: inline-flex;
`

const CarouselSlide = ({ children, ...restProps }: Props) => {
  return <Wrapper {...restProps}>{children}</Wrapper>
}

export default CarouselSlide
