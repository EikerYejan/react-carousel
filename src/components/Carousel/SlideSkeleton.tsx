import styled, { keyframes } from 'styled-components'
import { themeColor } from '../../theme'

const opacitySkeleton = keyframes`
    0% {
      opacity: 1;
    }
    50% {
      opacity: 0.7;
    }
    100% {
      opacity: 1;
    }
`

export default styled.div`
  width: 100%;
  height: 100%;
  background: ${themeColor('dark')};
  animation: ${opacitySkeleton} 1s linear infinite forwards;
  position: relative;
`
