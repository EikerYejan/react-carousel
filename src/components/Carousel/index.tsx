import React, { useEffect, useState } from 'react'
import styled from 'styled-components'
import { themeColor } from '../../theme'

export type CarouselOptions = {
  /**
   * Distance between slides in px.
   * @default 25
   */
  spaceBetween?: number

  /**
   * Duration of transition in seconds.
   * @default 0.25
   */
  transitionDuration?: number

  /**
   * The initial slide to display
   * @default 0
   */
  initialActiveIndex?: number

  /**
   *  Number of slides per view (slides visible at the same time on slider's container)
   * @default 1
   */
  slidesPerView?: number

  /**
   * Show navigation arrows
   * @default true
   */
  navigation?: boolean

  /**
   * Show pagination bullets
   * @default false
   */
  pagination?: boolean

  /**
   * Callback function to be called when the active slide changes
   * @param {number} index
   * @returns {void} void
   */
  onChangeSlide?: (index: number) => void
}

type Props = CarouselOptions & {
  children: React.ReactNode[]
  className?: string
}

type StyleProps = Pick<CarouselOptions, 'spaceBetween' | 'transitionDuration'>

const Wrapper = styled.div<StyleProps>`
  width: auto;
  overflow: hidden;
  position: relative;

  .carousel- {
    &inner {
      transition: transform ${({ transitionDuration }) => `${transitionDuration}s`};
      display: flex;
      flex-wrap: nowrap;
      gap: ${({ spaceBetween }) => `${spaceBetween}px`};
    }

    &navigation {
      background: none;
      border: none;
      outline: none;
      cursor: pointer;
      font-size: 18px;
      font-weight: bold;
      position: absolute;
      z-index: 1;
      color: ${themeColor('white')};
      width: 45px;
      height: 45px;
      top: calc(50% - 45px);
      padding: 0;
      transition: 0.5s;

      img {
        width: 100%;
        height: 100%;
        object-fit: cover;
        filter: invert(1);
      }

      &.next {
        right: 30px;
      }

      &.prev {
        left: 30px;
      }

      &:disabled {
        opacity: 0.35;
        cursor: not-allowed;
      }
    }

    &pagination {
      display: flex;
      justify-content: center;
      align-items: center;
      margin: 25px 0;
      gap: 5px;

      &-bullet {
        font-size: 0;
        width: 14px;
        min-width: 14px;
        min-height: 14px;
        height: 14px;
        display: block;
        padding: 0;
        background-color: ${themeColor('primary')};
        border: none;
        border-radius: 50%;
        opacity: 0.2;
        transition: 0.5s;
        cursor: pointer;
        flex-wrap: wrap;

        &.is-active {
          opacity: 1;
        }
      }
    }
  }
`

const Carousel = ({
  children,
  initialActiveIndex,
  slidesPerView = 1,
  spaceBetween = 25,
  transitionDuration = 0.25,
  navigation = true,
  pagination = false,
  className = '',
  onChangeSlide,
}: Props) => {
  const [activeIndex, setActiveIndex] = useState(initialActiveIndex ?? 0)
  const slidesCount = Math.ceil(React.Children.count(children) / slidesPerView) ?? 0

  const updateIndex = (i: number) => {
    let newIndex = i

    if (i < 0) newIndex = 0
    else if (newIndex >= slidesCount) newIndex = slidesCount - 1

    setActiveIndex(newIndex)
  }

  const prevButtonIsDisabled = activeIndex <= 0
  const nextButtonIsDisabled = activeIndex >= slidesCount - 1

  const styleProps = { spaceBetween, transitionDuration, className }
  const translateStyles = `translate3d(calc(-${activeIndex * 100}% - ${
    spaceBetween * activeIndex
  }px), 0, 0)`
  const slidesGap = spaceBetween * (slidesPerView - 1)
  const slideWidth = `calc((100vw - ${slidesGap}px) / ${slidesPerView})`

  useEffect(() => {
    if (initialActiveIndex) {
      if (initialActiveIndex >= slidesCount) setActiveIndex(slidesCount - 1)
      else setActiveIndex(initialActiveIndex)
    }
  }, [initialActiveIndex, slidesCount])

  useEffect(() => {
    if (onChangeSlide) onChangeSlide(activeIndex)
  }, [activeIndex, onChangeSlide])

  return (
    <Wrapper {...styleProps}>
      <div
        style={{ transform: translateStyles }}
        className="carousel-inner"
        data-testid="carousel-inner"
      >
        {React.Children.map(children, (child) => {
          return React.cloneElement(child as React.ReactElement, {
            style: { minWidth: slideWidth, width: slideWidth, maxWidth: slideWidth },
          })
        })}
      </div>
      {pagination && (
        <div className="carousel-pagination">
          {Array(slidesCount)
            .fill(0)
            .map((_, i) => (
              <button
                type="button"
                className={`carousel-pagination-bullet ${i === activeIndex ? 'is-active' : ''}`}
                onClick={() => updateIndex(i)}
                key={`bullet-${i}`}
              >
                {i + 1}
              </button>
            ))}
        </div>
      )}
      {navigation && (
        <>
          <button
            data-testid="navigation-prev"
            className="carousel-navigation prev"
            disabled={prevButtonIsDisabled}
            onClick={() => (!prevButtonIsDisabled ? updateIndex(activeIndex - 1) : null)}
            title="Previous"
            type="button"
          >
            <img src="/images/chevron-back.svg" alt="prev" />
          </button>
          <button
            data-testid="navigation-next"
            className="carousel-navigation next"
            disabled={nextButtonIsDisabled}
            onClick={() => updateIndex(activeIndex + 1)}
            title="Next"
            type="button"
          >
            <img src="/images/chevron-forward.svg" alt="next" />
          </button>
        </>
      )}
    </Wrapper>
  )
}

export default Carousel
