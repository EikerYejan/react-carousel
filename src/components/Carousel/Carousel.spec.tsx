import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Carousel from '.'
import CarouselSlide from './Slide'

const mockData = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10']

describe('Carousel', () => {
  it('Should render', () => {
    const { baseElement } = render(
      <Carousel>
        {mockData.map((i) => (
          <CarouselSlide key={i}>{`Slide ${i}`}</CarouselSlide>
        ))}
      </Carousel>,
    )

    expect(baseElement).toMatchSnapshot()
    expect(baseElement).toBeInTheDocument()
  })

  it('Should navigate with the arrows', () => {
    const { getByTestId } = render(
      <Carousel pagination>
        {mockData.map((i) => (
          <CarouselSlide key={i}>{`Slide ${i}`}</CarouselSlide>
        ))}
      </Carousel>,
    )

    const prevButton = getByTestId('navigation-prev')
    const nextButton = getByTestId('navigation-next')

    expect(prevButton).toBeInTheDocument()
    expect(nextButton).toBeInTheDocument()

    userEvent.click(nextButton)

    expect(getByTestId('carousel-inner')).toHaveStyle(
      `transform: translate3d(calc(-${1 * 100}% - ${25 * 1}px), 0, 0)`,
    )

    userEvent.click(prevButton)

    expect(getByTestId('carousel-inner')).toHaveStyle(
      `transform: translate3d(calc(-0% - 0px), 0, 0)`,
    )
  })
})
