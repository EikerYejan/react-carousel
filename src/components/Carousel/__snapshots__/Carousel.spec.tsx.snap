// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Should render 1`] = `
.c0 {
  width: auto;
  overflow: hidden;
  position: relative;
}

.c0 .carousel-inner {
  -webkit-transition: -webkit-transform 0.25s;
  -webkit-transition: transform 0.25s;
  transition: transform 0.25s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  gap: 25px;
}

.c0 .carousel-navigation {
  background: none;
  border: none;
  outline: none;
  cursor: pointer;
  font-size: 18px;
  font-weight: bold;
  position: absolute;
  z-index: 1;
  width: 45px;
  height: 45px;
  top: calc(50% - 45px);
  padding: 0;
  -webkit-transition: 0.5s;
  transition: 0.5s;
}

.c0 .carousel-navigation img {
  width: 100%;
  height: 100%;
  object-fit: cover;
  -webkit-filter: invert(1);
  filter: invert(1);
}

.c0 .carousel-navigation.next {
  right: 30px;
}

.c0 .carousel-navigation.prev {
  left: 30px;
}

.c0 .carousel-navigation:disabled {
  opacity: 0.35;
  cursor: not-allowed;
}

.c0 .carousel-pagination {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 25px 0;
  gap: 5px;
}

.c0 .carousel-pagination-bullet {
  font-size: 0;
  width: 14px;
  height: 14px;
  display: block;
  padding: 0;
  border: none;
  border-radius: 50%;
  opacity: 0.2;
  -webkit-transition: 0.5s;
  transition: 0.5s;
  cursor: pointer;
}

.c0 .carousel-pagination-bullet.is-active {
  opacity: 1;
}

.c1 {
  height: 100%;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

<body>
  <div>
    <div
      class="c0"
    >
      <div
        class="carousel-inner"
        data-testid="carousel-inner"
        style="transform: translate3d(calc(-0% - 0px), 0, 0);"
      >
        <div
          class="c1"
          style="min-width: calc((100vw - 0px) / 1); max-width: calc((100vw - 0px) / 1);"
        >
          Slide 1
        </div>
        <div
          class="c1"
          style="min-width: calc((100vw - 0px) / 1); max-width: calc((100vw - 0px) / 1);"
        >
          Slide 2
        </div>
        <div
          class="c1"
          style="min-width: calc((100vw - 0px) / 1); max-width: calc((100vw - 0px) / 1);"
        >
          Slide 3
        </div>
        <div
          class="c1"
          style="min-width: calc((100vw - 0px) / 1); max-width: calc((100vw - 0px) / 1);"
        >
          Slide 4
        </div>
        <div
          class="c1"
          style="min-width: calc((100vw - 0px) / 1); max-width: calc((100vw - 0px) / 1);"
        >
          Slide 5
        </div>
        <div
          class="c1"
          style="min-width: calc((100vw - 0px) / 1); max-width: calc((100vw - 0px) / 1);"
        >
          Slide 6
        </div>
        <div
          class="c1"
          style="min-width: calc((100vw - 0px) / 1); max-width: calc((100vw - 0px) / 1);"
        >
          Slide 7
        </div>
        <div
          class="c1"
          style="min-width: calc((100vw - 0px) / 1); max-width: calc((100vw - 0px) / 1);"
        >
          Slide 8
        </div>
        <div
          class="c1"
          style="min-width: calc((100vw - 0px) / 1); max-width: calc((100vw - 0px) / 1);"
        >
          Slide 9
        </div>
        <div
          class="c1"
          style="min-width: calc((100vw - 0px) / 1); max-width: calc((100vw - 0px) / 1);"
        >
          Slide 10
        </div>
      </div>
      <button
        class="carousel-navigation prev"
        data-testid="navigation-prev"
        disabled=""
        title="Previous"
        type="button"
      >
        <img
          alt="prev"
          src="/images/chevron-back.svg"
        />
      </button>
      <button
        class="carousel-navigation next"
        data-testid="navigation-next"
        title="Next"
        type="button"
      >
        <img
          alt="next"
          src="/images/chevron-forward.svg"
        />
      </button>
    </div>
  </div>
</body>
`;
